### ISSUE TITLE
Formatting of search results is poor

### ISSUE THREAD

**Issue by esafak** [nim-lang/nimforum#331]
Surely this is not a reasonable result for [static statement](https://forum.nim-lang.org/search?q=static%20statement):

![Screen Shot 2022-09-28 at 21 12 47](https://user-images.githubusercontent.com/3928300/192937206-7ae88240-1caf-45d2-92c0-a97a0a8d5c2d.png)

The code block is broken, and a random sentence is printed in a huge font while the title's weight is weaker than the author's.

**Comment by PhilippMDoerner** [comment]
I just tried to take a look at this and received this output in the forum:

![image](https://user-images.githubusercontent.com/65814317/214308817-ba1f6eac-9f45-4015-aba7-92c3e557b0ab.png)

I'm not 100%, but was the issue maybe fixed in the meantime and can thus be closed? Or am I overlooking anything?

**Comment by pietroppeter** [comment]
it does indeed seems it is fixed but looking at past commits it is unclear what fixed it (I do remember the issue in the past, do not remember recent occurences). and unclear whether or not the forum has been recently deployed. https://github.com/nim-lang/nimforum/issues/317 would be useful :)

**Comment by ringabout** [comment]
https://forum.nim-lang.org/t/7969#50770 seems to be broken: "Couldn't render historic post in #50770"

**Comment by pietroppeter** [comment]
mmh, if this screenshot is to be trusted we do have a last deploy last october: 
![image](https://user-images.githubusercontent.com/4997234/214329124-058fedc3-c394-4865-948c-ab47c38c7314.png)

**Comment by pietroppeter** [comment]
> https://forum.nim-lang.org/t/7969#50770 seems to be broken: "Couldn't render historic post in #50770"

this is related to this (still open) issue: https://github.com/nim-lang/nimforum/issues/330 

I think there are some cases still not working

**Comment by jyapayne** [comment]
@pietroppeter Search the forum for `option.sepBy`. You should get this:

<img width="1085" alt="image" src="https://user-images.githubusercontent.com/2645657/214337352-086756ff-469c-49c5-bd3a-f7d17b1c99fc.png">

The issue is only present when the code block gets cut off and doesn't have an ending and/or beginning ```

**Comment by PhilippMDoerner** [comment]
I can't seem to replicate when I run this locally.

I pretty much just copy pasted jack's post above to make a thread out of it, but no matter what I search, this is the kind of stuff I get:

![image](https://user-images.githubusercontent.com/65814317/214345080-9d37e4b2-6cbf-4c8a-991e-3bc39c452ba9.png)
![image](https://user-images.githubusercontent.com/65814317/214345241-1dc4a718-10b0-431d-9264-c9e8a07f45e5.png)
![image](https://user-images.githubusercontent.com/65814317/214345280-a23efe40-c78e-4218-9d31-33656996e88d.png)

This is with the forum compiled under 1.9.1, which shouldn't make a difference here.

**Comment by PhilippMDoerner** [comment]
Ohhhhh it's **explicitly the comment** that is making an issue here.
The combination of the comment being `##` (which means headline)  and no indication that this should be a codeblock being present.

**Comment by jyapayne** [comment]
@PhilippMDoerner You are correct. This issue would likely be fairly difficult to solve based on the reasons you mentioned. With the current structure of the code, I'm also not sure how to approach it. FTS with the `snippet` extension only returns the partial text from each match, so you'd probably need to get the full text from sqlite to compare so you'd be able to see where the start/end code block markers are. An ugly problem indeed. 

I was curious how Discourse solves this, and it looks like they just strip all formatting and render the plain text: https://discourse.codecombat.com/search?q=penPositions%20%3D. Maybe we could do that as well.

**Comment by pietroppeter** [comment]
> they just strip all formatting and render the plain text

It seems a very reasonable thing to do indeed. Maybe highlighting (bold?) the search term as they do

**Comment by PhilippMDoerner** [comment]
> > they just strip all formatting and render the plain text
> 
> It seems a very reasonable thing to do indeed. Maybe highlighting (bold?) the search term as they do

That actually runs into problems for exactly the example we were looking at.
Say the comment looks like this:

` ## [<p1> (<p2> <p1>)*]`

And you have a searchquery of `<p1>`.
Now naturally you can just go and do a "replace" on the HTML string to insert `<strong> <p1> </strong>` for any hit to`<p1>` in the string.
The issue comes in rendering. 
How do you render this int he frontend correctly now, because you want the `<strong>` tags to render as HTML, while you want `<p1>` to render as text?
If you try to render this as HTML, you'll get the ` <p1>` bold but you also get <p1> interpreted to be rendered as HTML and it starts disappearing.

Demonstration:
![image](https://user-images.githubusercontent.com/65814317/214593856-84fa4b1f-a46b-4728-b868-847ade125654.png)

As you can see, there is no <p1> to be found anywhere.

Now we can of course circumvent that, we can say to just render the text as is. But that means it'll also render the `<strong>` tags for text instead of as HTML.

Demonstration:
![image](https://user-images.githubusercontent.com/65814317/214594542-8d29bee5-5a0c-42b7-8be3-dc6c800cdea9.png)

So either we just render plaintext, or we accept that any HTML-strings in the content itself will not be rendered as text.

As an additional issue though, I haven't found an easy way in karax so far to retain the whitespacing, so it all looks like the unreadable messy textblobs you see above.

**Comment by jyapayne** [comment]
@PhilippMDoerner If you want to render a search of `<p1>`, you will have to escape the characters that HTML considers special characters. See [here](https://mateam.net/html-escape-characters/) for a complete table of conversions. I believe you should only have to escape `&`, `<`, `>`, `"`, and `'`.

**Comment by jyapayne** [comment]
So in your case, you would replace `<p1>` by `<strong> &lt;p1&gt; </strong>`

**Comment by PhilippMDoerner** [comment]
It does turn out we're overcomplicating things.
I just throw together a bit of a solution for multiple issues:

1) The snippets that get rendered (using SQLITE FTS4-SQL feature called SNIPPET) sometimes fail to render properly. You instead get to actually see the `**` that should get rendered into strong tags, because there's no whitespace after the match. I added those, that get rid of an entire class of rendering issues
2) To prevent `## ` and `# ` (the leading whitespace is important) to be rendered into a headline, we might as well do a simple replace of `# ` in general to `\\# `. **Most** of the time that fixes the issue, sometimes that leads to `\#` being rendered instead of `#`. Which I can live with.

Mysteriously that even takes care of the HTML issue

![image](https://user-images.githubusercontent.com/65814317/214867733-0ed45e0a-012e-44ec-ad6a-71f97f3c838e.png)

**Comment by jyapayne** [comment]
That fixes the bolded text issue, but now you have more issues because the markdown is parsing `proc sepBy[T, T2](p1: Parser[T], p2: Parser[P2]): Parser[seq[T]] =` as `proc sepBy T, T2: Parser[seq[T]]`. You're losing information in the text because markdown sees `[X](Y)` and parses it as a link `Y` with text `X`. I still think you're better off using the plaintext and  escaping the HTML special characters as I mentioned above. Otherwise you're going to have to keep finding hacks to workaround markdown parsing issues.

**Comment by PhilippMDoerner** [comment]
The alternative is plain-text and hacks that work around making things sometimes bold and sometimes not, we're running into hacks either way.

**Comment by jyapayne** [comment]
But they aren't hacks. HTML has special characters that need to be escaped if you want to display them normally. Replacing `&`, `<`, `>`, `"`, and `'` with their respective escape codes is how you insert text into an HTML element in JS. 

I get the formatting will be cleaner, but that isn't really an issue when you can just click into the post and read the full code. There is a good reason that Discourse doesn't render text with formatting. It's tricky business and broken text gets rendered like it currently does in the Nim forum.

Also, if you start escaping Markdown characters, the formatting in the regular, non-code text will be broken and won't display headers and links correctly.
